### PR TITLE
Fix commenting on non-utf8 encoded files

### PR DESCRIPTION
**Pull request by zeripath**
The original implementation of models.Comment.Patch maps this to an SQL TEXT field. This works fine when the line being commented on is UTF-8 however, if there are non-UTF-8 characters this fails on MySQL and PostgreSQL databases.

This PR changes this field to quote it if necessary.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
Finally! This PR now replicates the errors seen in #7270

File: integrations/integration_test.go [hunk 1]
```diff
@@ -160,8 +160,12 @@ func initIntegrationTest() {
 
 	switch {
 	case setting.Database.UseMySQL:
-		db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@tcp(%s)/",
-			setting.Database.User, setting.Database.Passwd, setting.Database.Host))
+		connType := "tcp"
+		if len(setting.Database.Host) > 0 && setting.Database.Host[0] == '/' { // looks like a unix socket
```
Comment (lunny) [inline integrations/integration_test.go]: The change is unrelated with this PR.

File: integrations/migration-test/migration_test.go [hunk 2]
```diff
@@ -142,8 +142,12 @@ func restoreOldDB(t *testing.T, version string) bool {
 		db.Close()
 
 	case setting.Database.UseMySQL:
-		db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@tcp(%s)/",
-			setting.Database.User, setting.Database.Passwd, setting.Database.Host))
+		connType := "tcp"
```
Comment (lunny) [inline integrations/migration-test/migration_test.go]: The same

File: models/migrations/v143.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	"xorm.io/xorm"
+	"xorm.io/xorm/schemas"
+)
+
+func changePatchToBlob(x *xorm.Engine) error {
+	switch x.Dialect().URI().DBType {
+	case schemas.MYSQL:
+		_, err := x.Exec("ALTER TABLE comment MODIFY COLUMN patch BLOB")
+		return err
+	case schemas.ORACLE:
```
Comment (lunny) [inline models/migrations/v143.go]: We haven't support `Oracle` yet. :)

File: models/migrations/v143.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+			return err
+		}
+		defer sess.Close()
+		if _, err := sess.Exec("SELECT id, patch INTO #tempcomment FROM comment"); err != nil {
+			return fmt.Errorf("Unable to create temporary table: %v", err)
+		}
+		if _, err := sess.Exec("ALTER TABLE comment DROP COLUMN patch"); err != nil {
+			return fmt.Errorf("DROP COLUMN patch: %v", err)
+		}
+		if _, err := sess.Exec("ALTER TABLE comment ADD COLUMN patch VARBINARY(8000);"); err != nil {
```
Comment (lafriks) [inline models/migrations/v143.go]:
```suggestion
		if _, err := sess.Exec("ALTER TABLE comment ADD COLUMN patch VARBINARY(MAX);"); err != nil {
```

File: models/issue_comment.go [hunk 5]
```diff
@@ -143,7 +143,7 @@ type Comment struct {
 	RenderedContent string `xorm:"-"`
 
 	// Path represents the 4 lines of code cemented by this comment
-	Patch string `xorm:"TEXT"`
+	Patch string `xorm:"BLOB(8000)"`
```
Comment (lafriks) [inline models/issue_comment.go]:
```suggestion
	Patch string `xorm:"BLOB"`
```

File: models/migrations/v143.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+			return fmt.Errorf("ADD COLUMN patch: %v", err)
+		}
+		if _, err := sess.Exec("UPDATE comment SET patch = CONVERT(VARBINARY(8000), tmp.patch) FROM comment INNER JOIN #tempcomment ON comment.id = #tempcomment.id "); err != nil {
```
Comment (lafriks) [inline models/migrations/v143.go]:
```suggestion
		if _, err := sess.Exec("UPDATE comment SET patch = CONVERT(VARBINARY(MAX), tmp.patch) FROM comment INNER JOIN #tempcomment ON comment.id = #tempcomment.id "); err != nil {
```

File: models/issue_comment.go (hunk 5 above)
Comment (lafriks) [reply models/issue_comment.go]: This depends on https://gitea.com/xorm/xorm/issues/1716 for new instances

**Comment by zeripath** [comment]
Hmm... seems like Blob won't work without adding functions to convert between blob and text.

In which case we may as well keep it as TEXT and just escape the bad chars.

File: models/issue_comment.go (hunk 5 above)
Comment (lafriks) [inline models/issue_comment.go]:
```suggestion
	Patch byte[] `xorm:"BLOB(8000)"`
```

**Comment by lafriks** [comment]
You should change struct Patch to byte array instead of string

**Comment by zeripath** [comment]
I think the simplest thing is to shadow the Patch column and quote/unquote as necessary.

**Comment by lafriks** [comment]
Can't we just convert to UTF-8 when saving if it is not valid UTF-8 like we do it for displaying with detect encoding?

**Review by lunny — APPROVED** [review]

**Comment by zeripath** [comment]
So I've thought about that but:

a) We don't have a long enough piece of text to detect the encoding and we can't assume that we will always be enable to detect the encoding.
b) How do you demonstrate changes of encoding? It's entirely possible that one side of the change has a different encoding to the other.
c) That breaks the patch - admittedly we don't use the patch at present except for demonstration but I am working on something at present that might use the patch.

Keeping the broken bytes is really better as we can use them to improve our rendering in future.

**Review by 6543 — APPROVED** [review]
to fix it this is totaly ok - we should just konsiddert to refactor&improve this at some point

**Comment by 6543** [comment]
🚀

**Comment by 6543** [comment]
I'll send a backport
